### PR TITLE
Renamed `Crc64` to `Crc64Hash` and change it to derive from the `Azure::Core::Cryptography::Hash` class.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Follow-up from https://github.com/Azure/azure-sdk-for-cpp/pull/1632

Also partially addresses https://github.com/Azure/azure-sdk-for-cpp/issues/1619 for Crc64 as well, similar to Md5 (topic 1 and 2).

I will leave it up to @Jinming-Hu and storage folks to help decide whether they want this change in this release, or wait for the next one.

### CODE REVIEW

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:13]: @JeffreyRichter, @RickWinter should this be in a Cryptography sub-namespace, i.e. `Azure::Storage::Cryptography`?

Comment (RickWinter) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:13]: Yes, I suggested Crypto b/c that is what some other languages use.  I'm open to any good name.  I suspect the set grows and thus my push for them to be in a namespace.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (JeffreyRichter) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:15]: I think Jinming has a reason why this class will NOT derive from Core's Hash.

**Review by JeffreyRichter — APPROVED** [review]

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:15]: >  If it's just because it needs Concatenate but it ALSO works/is useful without Concatenate, then deriving seems good to me.

I think that was his reasoning, but having a concrete class add additional specific APIs is fine, from https://github.com/Azure/azure-sdk-for-cpp/pull/1632#issuecomment-775643215, unless there are other reasons.
> From @Jinming-Hu:
> Plus, I'm not going to make Crc64 class inherit from this hash base. As Crc64 supports concatenate operation, while your base class doesn't.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:13]: Right, @RickWinter, that makes sense for `Azure::Core::Cryptography` and the base `Hash` class (which we already did). Does that apply to services that derive from the base as well, such as Crc64 in Storage?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:15]: At least one benefit for deriving like this, is that we get consistent and well-defined/documented behavior for the edge cases for free such as what happens when you call `Final()` multiple times or `Append()` after a call to `Final()`. The base class enforces the right behavior.

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]
I'm gonna take a look today, wait for storage's approval

**Comment by ahsonkhan** [comment]
> I'm gonna take a look today, wait for storage's approval

I wouldn't merge without it 😊

File: sdk/storage/azure-storage-blobs/test/page_blob_client_test.cpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/test/page_blob_client_test.cpp:286]: Nit: can we add a static method `Hash`?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:31]: Can we keep this?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:18]: need override destructor

Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:18]: What will we do in the destructor? It wasn't there in the previous implementation.

File: sdk/storage/azure-storage-blobs/test/page_blob_client_test.cpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-blobs/test/page_blob_client_test.cpp:286]: We can add it back later if we get customer feedback, and @JeffreyRichter recommended I remove it for now (same with `Md5Hash`).

How commonly do you think customers would call the `single-shot` hashing API, rather than the streaming one?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:31]: Will folks typically try to compute Crc64 hash for text string data? Or is it similar to other hashing (like md5) which is meant for files/blobs that tend to binary/bytes?

These are convenience methods which we can easily add later, and so can the end-customer if they really want to.

It's not too bad atm:
```C++
CrcHash instance;
auto hash = instance.Final(data, length);
```
versus
```C++
auto hash = CrcHash::Hash(data, length);
```

File: sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:18]: nothing, just 

```C++
Crc64Hash() override = default;
```

to enforce the base class has a virtual destructor.

Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:18]: Seems odd to do. The base class already has a virtual destructor, why would the concrete derived class need to enforce such a thing?

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:18]: Let's say some customer cares a lot about data integrity and always wants to use Md5 or Crc64 or whatever checksum servicer supports. He also wants to be able to switch different hash algorithm with an option, possible value of which can be MD5, CRC64, and others. To avoid duplicate logic, he will use a pointer to the hash base class `HashBase*` and instantiate a concrete hash instance `Md5Hash` or `Crc64Hash` and assign the point to `HashBase*`. Now he can just call `HashBase->Append` to append data and `HashBase->Final` to get final hash value, without caring about the hash algorithm. This helps a lot in reducing code redundancy. Once the operation is done, he will deallocate the hash instance by `delete HashBase`. If the destructor of `HashBase` is not virtual, it will cause potential resource leak because the destructor of concrete hash class isn't executed. Now to avoid this, we will use `override` on the destructor of concrete class. It's always a good habit if some object is possible to be deleted via base pointer.

> The base class already has a virtual destructor, why would the concrete derived class need to enforce such a thing? What's the benefit?

same question goes for [OnFinal](https://github.com/Azure/azure-sdk-for-cpp/blob/2d4d766c6d454bfbd2a3627c5a3d8a4d81196480/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L136) and [OnAppend](https://github.com/Azure/azure-sdk-for-cpp/blob/2d4d766c6d454bfbd2a3627c5a3d8a4d81196480/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L146). We don't need `override` to be grammatically correct.

Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:18]: > If the destructor of `HashBase` is not virtual, it will cause potential resource leak because the destructor of concrete hash class isn't executed.

But it is virtual, so why safeguard against something that is already true?

> same question goes for [OnFinal](https://github.com/Azure/azure-sdk-for-cpp/blob/2d4d766c6d454bfbd2a3627c5a3d8a4d81196480/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L136) and [OnAppend](https://github.com/Azure/azure-sdk-for-cpp/blob/2d4d766c6d454bfbd2a3627c5a3d8a4d81196480/sdk/core/azure-core/inc/azure/core/cryptography/hash.hpp#L146). We don't need `override` to be grammatically correct. I just think it's a good habit

But in this case, we are actually overriding the behavior/logic of those functions.

The `HashBase` already has an implementation of the virtual destructor as `default`, and if `Crc64Hash` is not changing the implementation or doing something meaningful, why override it prematurely? I agree that **IF** the destructor needs to be overridden, we'd use the `override` keyword. I am asking what's the benefit of overriding it today?

All that said, I am fine with adding it.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:18]: > But in this case, we are actually overriding the behavior/logic of those functions.

Without `override`, you're also overriding the behavior.

> The HashBase already has an implementation of the virtual destructor as default, and if Crc64Hash is not changing the implementation or doing something meaningful, why override it prematurely?

As you say, it's just a safeguard. Who's to say nobody will change the impl/interface of `HashBase` and forget to make the destructor virtual. If we use `override` in derived destructor, the compiler can remind us.

Comment (ahsonkhan) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/crypt.hpp:18]: > Without `override`, you're also overriding the behavior.

But currently I am not providing a destructor for `Crc64Hash` at all. How am I overriding anything? I understand **IF** I did provide one, the `override` keyword is helpful, but if I don't add a destructor at all, what's the issue?

Are you suggesting I **add** a destructor?

**Comment by ahsonkhan** [comment]
Looking for PR approval here to avoid making this storage breaking change after the Feb release.

**Review by Jinming-Hu — APPROVED** [review]

**Pull request merged**
